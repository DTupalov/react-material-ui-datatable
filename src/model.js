import { combine, createEvent, createStore, createStoreObject } from 'effector';
import {
  convertDataToFilterLists,
  filter,
  paginate,
  pipe,
  search,
  sort as sortFn,
} from './utils';

const createWatcher = (onChange, store) => name => value => {
  const {
    searchValue,
    showSearchBar,
    sort,
    filterValues,
    page,
    perPage,
    selectedData,
  } = store.getState();

  onChange({
    name,
    value,
    state: {
      searchValue,
      showSearchBar,
      sort,
      filterValues,
      page,
      perPage,
      selectedData,
    },
  });
};

const createModel = ({
  data: initialData = [],
  columns: initialColumns = [],
  showSearchBar: initialShowSearchBar = false,
  searchValue: initialSearchValue = '',
  sort: initialSort = { columnName: null, direction: 'ASC' },
  page: initialPage = 0,
  perPage: initialPerPage = 5,
  selectedData: initialSelectedData = [],
  filterValues: initialFilterValues = {},
}) => {
  /** data */
  const data = createStore(initialData);
  const changeData = createEvent();
  const handleDelete = createEvent();

  data
    .on(changeData, (_, nextData) => nextData)
    .on(handleDelete, (data, deletedData) =>
      data.filter(dataItem => !deletedData.includes(dataItem))
    );

  /** columns */
  const columns = createStore(initialColumns);
  const changeColumns = createEvent();

  columns.on(changeColumns, (_, nextColumns) => nextColumns);

  /** sort */
  const sort = createStore(initialSort);
  const handleSort = createEvent();

  sort.on(handleSort, (_, { columnName, direction = 'ASC' }) => ({
    columnName,
    direction,
  }));

  /** showSearchBar */
  const showSearchBar = createStore(initialShowSearchBar);
  const changeShowSearchBar = createEvent();
  const toggleSearchBar = createEvent();

  showSearchBar
    .on(toggleSearchBar, showSearchBar => !showSearchBar)
    .on(changeShowSearchBar, (_, nextShowSearchBar) => nextShowSearchBar);

  /** searchValue */
  const searchValue = createStore(initialSearchValue);
  const handleSearchValue = createEvent();

  searchValue
    .on(handleSearchValue, (_, nextSearchValue) => nextSearchValue)
    .reset(toggleSearchBar);

  /** selectedData */
  const selectedData = createStore(initialSelectedData);
  const toggleSelectRow = createEvent();
  const toggleSelectAll = createEvent();
  const handleSelect = createEvent();

  selectedData
    .on(handleSelect, (_, nextSelectedData) => nextSelectedData)
    .on(toggleSelectRow, (selectedData, selectedDataItem) => {
      const nextSelectedData = [...selectedData];
      const dataIndex = nextSelectedData.findIndex(
        data => data === selectedDataItem
      );
      if (dataIndex !== -1) {
        nextSelectedData.splice(dataIndex, 1);
      } else {
        nextSelectedData.push(selectedDataItem);
      }
      return nextSelectedData;
    })
    .on(toggleSelectAll, selectedData =>
      selectedData.length ? [] : [...data.getState()]
    )
    .on(handleDelete, (selectedData, deletedData) =>
      selectedData.filter(dataIndex => !deletedData.includes(dataIndex))
    );

  /** filterValues */
  const filterValues = createStore(initialFilterValues);
  const changeFilterValues = createEvent();
  const addFilter = createEvent();
  const removeFilter = createEvent();
  const resetFilter = createEvent();

  filterValues
    .on(addFilter, (filterValues, { columnName, value }) => ({
      ...filterValues,
      [columnName]: value,
    }))
    .on(removeFilter, (filterValues, { columnName }) =>
      Object.keys(filterValues).reduce((nextFilterValues, prevColumnName) => {
        if (prevColumnName !== columnName) {
          nextFilterValues[prevColumnName] = filterValues[prevColumnName];
        }
        return nextFilterValues;
      }, {})
    )
    .on(changeFilterValues, (_, nextFilterValues) => nextFilterValues)
    .reset(resetFilter);

  /** page */
  const page = createStore(initialPage);
  const changePage = createEvent();

  page.on(changePage, (_, nextPage) => nextPage);

  /** perPage */
  const perPage = createStore(initialPerPage);
  const changePerPage = createEvent();

  perPage.on(changePerPage, (_, nextPerPage) => nextPerPage);

  /** Computed Stores */
  const computedData = combine(
    data,
    columns,
    sort,
    searchValue,
    filterValues,
    (data, columns, sort, searchValue, filterValues) =>
      pipe(
        sortFn(sort.columnName, sort.direction),
        search(searchValue, columns),
        filter(filterValues, columns)
      )(data)
  );
  const filterLists = combine(computedData, columns, (computedData, columns) =>
    convertDataToFilterLists(computedData, columns)
  );
  const displayData = combine(
    computedData,
    page,
    perPage,
    (computedData, page, perPage) => paginate(page, perPage)(computedData)
  );

  /** Stores object */
  const store = createStoreObject({
    data,
    columns,
    sort,
    filterValues,
    searchValue,
    showSearchBar,
    selectedData,
    page,
    perPage,
    computedData,
    filterLists,
    displayData,
  });

  /** Actions object */
  const actions = {
    changeData,
    changeColumns,
    changeShowSearchBar,
    changeFilterValues,
    toggleSelectRow,
    toggleSelectAll,
    toggleSearchBar,
    handleSearchValue,
    handleSort,
    addFilter,
    removeFilter,
    resetFilter,
    changePage,
    changePerPage,
    handleSelect,
    handleDelete,
  };

  const subscribe = subscriber => {
    const createStoreWatcher = createWatcher(subscriber, store);

    const unwatchShowSearchBar = showSearchBar.watch(
      createStoreWatcher('showSearchBar')
    );
    const unwatchPage = page.watch(createStoreWatcher('page'));
    const unwatchPerPage = perPage.watch(createStoreWatcher('perPage'));
    const unwatchSelectedData = selectedData.watch(
      createStoreWatcher('selectedData')
    );
    const unwatchSearchValue = searchValue.watch(
      createStoreWatcher('searchValue')
    );
    const unwatchSort = sort.watch(createStoreWatcher('sort'));
    const unwatchFilterValues = filterValues.watch(
      createStoreWatcher('filterValues')
    );

    return () => {
      unwatchShowSearchBar();
      unwatchPage();
      unwatchPerPage();
      unwatchSelectedData();
      unwatchSearchValue();
      unwatchSort();
      unwatchFilterValues();
    };
  };

  return { store, actions, subscribe };
};

export default createModel;
