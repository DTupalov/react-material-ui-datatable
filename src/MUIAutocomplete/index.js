import MUIAutocomplete from './MUIAutocomplete';

export { MUIAutocomplete };
