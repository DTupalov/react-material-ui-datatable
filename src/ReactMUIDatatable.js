import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import React from 'react';
import ReactMUIDatatableBody from './ReactMUIDatatableBody';
import ReactMUIDatatableFooter from './ReactMUIDatatableFooter';
import ReactMUIDatatableHeader from './ReactMUIDatatableHeader';
import ReactMUIDatatableProvider, {
  ReactMUIDatatableConsumer,
} from './ReactMUIDatatableProvider';
import ReactMUIDatatableToolbar from './ReactMUIDatatableToolbar';
import ReactMUIDatatableToolbarFilterValues from './ReactMUIDatatableToolbarFilterValues';
import ReactMUIDatatableToolbarSelect from './ReactMUIDatatableToolbarSelect';

const ReactMUIDatatable = props => {
  return (
    <ReactMUIDatatableProvider {...props}>
      <ReactMUIDatatableConsumer>
        {({ selectedRows }) => (
          <Paper>
            {selectedRows.length ? (
              <ReactMUIDatatableToolbarSelect />
            ) : (
              <ReactMUIDatatableToolbar />
            )}
            <ReactMUIDatatableToolbarFilterValues />
            <Table>
              <ReactMUIDatatableHeader />
              <ReactMUIDatatableBody />
              <ReactMUIDatatableFooter />
            </Table>
          </Paper>
        )}
      </ReactMUIDatatableConsumer>
    </ReactMUIDatatableProvider>
  );
};

export default ReactMUIDatatable;
