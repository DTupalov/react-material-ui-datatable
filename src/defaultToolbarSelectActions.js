import { IconButton } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';
import React from 'react';

const defaultToolbarSelectActions = ({ selectedRows, handleDelete }) => {
  return (
    <IconButton
      onClick={() => {
        handleDelete(selectedRows);
      }}
    >
      <DeleteIcon />
    </IconButton>
  );
};

export default defaultToolbarSelectActions;
