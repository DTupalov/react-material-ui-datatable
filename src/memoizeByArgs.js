export const cache = new Map();

const hasCachedResult = (args, fn, cache) =>
  args.every((arg, index) => {
    const hasCachedValue = cache.has(arg);
    if (hasCachedValue) {
      cache = cache.get(arg);
    }

    return hasCachedValue;
  });

const getCachedResult = (args, fn, cache) =>
  args.reduce((nextCachedValue, arg) => nextCachedValue.get(arg), cache);

const setCachedResult = (args, fn, cache) => {
  const fnResult = fn.apply(null, args);

  args.reduce((cachedValue, arg, index, args) => {
    if (cachedValue.has(arg)) return cachedValue.get(arg);

    let nextCachedValue;
    const isLastArg = index === args.length - 1;

    if (isLastArg) {
      nextCachedValue = fnResult;
    } else {
      nextCachedValue = new Map();
    }

    cachedValue.set(arg, nextCachedValue);

    return nextCachedValue;
  }, cache);

  return fnResult;
};

const fillArgsByFnLength = (args, fnLength) =>
  new Array(fnLength).fill(undefined).map((v, i) => args[i]);

const cachedResult = (args, fn, cache) =>
  hasCachedResult(args, fn, cache)
    ? getCachedResult(args, fn, cache)
    : setCachedResult(args, fn, cache);

export default fn => {
  const cachedFn = cache.get(fn) || (cache.set(fn, new Map()) && cache.get(fn));

  return (...args) => {
    const filledArgs = fillArgsByFnLength(args, fn.length);
    const result = cachedResult(filledArgs, fn, cachedFn);
    return result;
  };
};
