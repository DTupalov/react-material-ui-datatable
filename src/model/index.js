import createModel from './createModel';

export { createModel };
