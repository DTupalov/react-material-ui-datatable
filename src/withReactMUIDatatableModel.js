import memoize from 'lodash/memoize';
import compose from 'recompose/compose';
import defaultProps from 'recompose/defaultProps';
import withHandlers from 'recompose/withHandlers';
import withProps from 'recompose/withProps';
import withStateHandlers from 'recompose/withStateHandlers';
import defaultToolbarSelectActions from './defaultToolbarSelectActions';
import {
  addMetaRawIndexToData,
  completeColumnsWithOptions,
  convertColumnsToFilterValues,
  convertDataToFilterLists,
  filter,
  metaSymbol,
  paginate,
  pipe,
  search,
  sort,
} from './utils';

export const mapDatatableProps = props => ({
  title: props.title,
  columns: props.columns,
  search: props.search,
  sort: props.sort,
  filterValues: props.filterValues,
  page: props.page,
  perPage: props.perPage,
  selectedRows: props.selectedRows,
  toolbarSelectActions: props.toolbarSelectActions,
  page: props.page,
  perPage: props.perPage,
  perPageOption: props.perPageOption,
  selectable: props.selectable,
  filterable: props.filterable,
  searchable: props.searchable,
  localization: props.localization,
});

export const mapDatatableCalculatedProps = props => ({
  computedData: props.computedData,
  filterLists: props.filterLists,
  displayData: props.displayData,
});

export const mapDatatableHandlers = props => ({
  toggleSelectRow: props.toggleSelectRow,
  toggleSelectAll: props.toggleSelectAll,
  toggleSearchBar: props.toggleSearchBar,
  handleSearchValue: props.handleSearchValue,
  handleSort: props.handleSort,
  addFilter: props.addFilter,
  removeFilter: props.removeFilter,
  resetFilter: props.resetFilter,
  changePage: props.changePage,
  changePerPage: props.changePerPage,
  handleSelect: props.handleSelect,
  handleDelete: props.handleDelete,
});

export default compose(
  defaultProps({
    data: [],
    columns: [],
    title: '',
    search: {
      showSearchBar: false,
      value: '',
    },
    sort: { columnName: null, direction: 'ASC' },
    toolbarSelectActions: defaultToolbarSelectActions,
    page: 0,
    perPage: 5,
    perPageOption: [5, 10, 15],
    selectedRows: [],
    selectable: false,
    filterable: false,
    searchable: false,
    localization: {
      toolbar: {
        searchAction: 'Search',
        filterAction: 'Filters',
        closeSearch: 'Close search',
      },
      filterLists: {
        title: 'Filter',
        allOption: 'All',
        reset: 'Reset',
      },
      toolbarSelect: {
        selectedRows: count => `${count} row(s) selected`,
      },
      pagination: {
        rowsPerPage: 'Rows per page',
        displayedRows: ({ from, to, count }) => `${from}-${to} of ${count}`,
      },
    },
  }),
  withProps(props => ({
    data: addMetaRawIndexToData(props.data),
    columns: completeColumnsWithOptions(props.columns),
    filterValues: convertColumnsToFilterValues(props.columns),
  })),
  withStateHandlers(
    props => ({
      data: props.data,
      search: props.search,
      sort: props.sort,
      filterValues: props.filterValues,
      page: props.page,
      perPage: props.perPage,
      selectedRows: props.selectedRows,
    }),
    {
      setData: () => data => ({ data }),
      setSearch: () => search => ({ search }),
      setSort: () => sort => ({ sort }),
      setFilterValues: () => filterValues => ({ filterValues }),
      setPage: () => page => ({ page }),
      setPerPage: () => perPage => ({ perPage }),
      setSelectedRows: () => selectedRows => ({ selectedRows }),
    }
  ),
  withProps(
    memoize(
      props => {
        const computedData = pipe(
          sort({
            columnName: props.sort.columnName,
            direction: props.sort.direction,
          }),
          search({ value: props.search.value, columns: props.columns }),
          filter({ values: props.filterValues, columns: props.columns })
        )(props.data);

        const filterLists = convertDataToFilterLists({
          data: computedData,
          columns: props.columns,
        });

        const displayData = paginate({
          page: props.page,
          perPage: props.perPage,
        })(computedData);

        return {
          computedData,
          filterLists,
          displayData,
        };
      },
      props =>
        JSON.stringify([
          props.data,
          props.columns,
          props.sort.columnName,
          props.sort.direction,
          props.search.value,
          props.filterValues,
          props.page,
          props.perPage,
        ])
    )
  ),
  withHandlers({
    toggleSelectRow: props => rawIndex => {
      const nextSelectedRows = [...props.selectedRows];
      const indexOfRow = nextSelectedRows.indexOf(rawIndex);

      if (indexOfRow !== -1) {
        nextSelectedRows.splice(indexOfRow, 1);
      } else {
        nextSelectedRows.push(rawIndex);
      }

      props.setSelectedRows(nextSelectedRows);
    },
    toggleSelectAll: props => () => {
      let nextSelectedRows = [];
      if (!props.selectedRows.length) {
        nextSelectedRows = props.data.map(row => row[metaSymbol].rawIndex);
      }

      props.setSelectedRows(nextSelectedRows);
    },
    toggleSearchBar: props => () =>
      props.setSearch({
        showSearchBar: !props.search.showSearchBar,
        value: '',
      }),
    handleSearchValue: props => value =>
      props.setSearch({
        showSearchBar: props.search.showSearchBar,
        value,
      }),
    handleSort: props => ({ columnName, direction = 'ASC' }) =>
      props.setSort({ columnName, direction }),
    addFilter: props => ({ columnName, value }) =>
      props.setFilterValues({
        ...props.filterValues,
        [columnName]: value,
      }),
    removeFilter: props => ({ columnName }) =>
      props.setFilterValues({
        ...props.filterValues,
        [columnName]: '',
      }),
    resetFilter: props => () =>
      props.setFilterValues(convertColumnsToFilterValues(props.columns)),
    changePage: props => page => props.setPage(page),
    changePerPage: props => count => props.setPerPage(count),
    handleSelect: props => selectedRows => props.setSelectedRows(selectedRows),
    handleDelete: props => selectedRows => {
      const nextData = props.data.filter(
        row => !selectedRows.includes(row[metaSymbol].rawIndex)
      );

      props.setData(nextData);
      props.setSelectedRows([]);
    },
  })
);
