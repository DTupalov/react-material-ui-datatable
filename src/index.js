import ReactMUIDatatable from './ReactMUIDatatable';
import { metaSymbol } from './utils';
import withReactMUIDatatableModel from './withReactMUIDatatableModel';

export { ReactMUIDatatable, withReactMUIDatatableModel, metaSymbol };
