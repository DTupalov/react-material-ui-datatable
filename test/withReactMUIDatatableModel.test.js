import cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import renderer from 'react-test-renderer';
import { default as withReactMUIDatatableModel } from '../src/withReactMUIDatatableModel';

let ReactMUIDatatable;
let RenderedMockedComponent;
let data;
let columns;
let onStateChanged;
class MockedComponent extends React.Component {
  render() {
    // console.log(this.props);
    return null;
  }
}

describe('withReactMUIDatatableModel', () => {
  beforeEach(() => {
    onStateChanged = jest.fn(event => {});

    data = [
      {
        name: 'Caz',
        age: 49,
        car: {
          make: 'Land Rover',
        },
      },
      {
        name: 'Dav',
        age: 69,
        car: {
          make: 'Mitsubishi',
        },
      },
      {
        name: 'Yule',
        age: 68,
        car: {
          make: 'Ford',
        },
      },
      {
        name: 'Fredra',
        age: 67,
        car: {
          make: 'GMC',
        },
      },
      {
        name: 'Ursula',
        age: 80,
        car: {
          make: 'BMW',
        },
      },
      {
        name: 'Ailsun',
        age: 41,
        car: {
          make: 'Honda',
        },
      },
    ];

    columns = [
      { name: 'name', label: 'Name' },
      { name: 'age', label: 'Age' },
      { name: 'car.make', label: 'Car make' },
    ];

    ReactMUIDatatable = withReactMUIDatatableModel(
      jest.fn(props => <MockedComponent {...props} />)
    );

    RenderedMockedComponent = renderer
      .create(
        <ReactMUIDatatable
          data={data}
          columns={columns}
          onStateChanged={onStateChanged}
        />
      )
      .root.findByType(MockedComponent);
  });

  it('should sort data by column name with dots in ASC direction', () => {
    RenderedMockedComponent.props.handleSort({ columnName: 'car.make' });

    const expectedData = [
      {
        name: 'Ursula',
        age: 80,
        car: {
          make: 'BMW',
        },
      },
      {
        name: 'Yule',
        age: 68,
        car: {
          make: 'Ford',
        },
      },
      {
        name: 'Fredra',
        age: 67,
        car: {
          make: 'GMC',
        },
      },
      {
        name: 'Ailsun',
        age: 41,
        car: {
          make: 'Honda',
        },
      },
      {
        name: 'Caz',
        age: 49,
        car: {
          make: 'Land Rover',
        },
      },
      {
        name: 'Dav',
        age: 69,
        car: {
          make: 'Mitsubishi',
        },
      },
    ];

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should sort data by column name with dots in DESC direction', () => {
    RenderedMockedComponent.props.handleSort({
      columnName: 'car.make',
      direction: 'DESC',
    });

    const expectedData = [
      {
        name: 'Dav',
        age: 69,
        car: {
          make: 'Mitsubishi',
        },
      },
      {
        name: 'Caz',
        age: 49,
        car: {
          make: 'Land Rover',
        },
      },
      {
        name: 'Ailsun',
        age: 41,
        car: {
          make: 'Honda',
        },
      },
      {
        name: 'Fredra',
        age: 67,
        car: {
          make: 'GMC',
        },
      },
      {
        name: 'Yule',
        age: 68,
        car: {
          make: 'Ford',
        },
      },
      {
        name: 'Ursula',
        age: 80,
        car: {
          make: 'BMW',
        },
      },
    ];

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should filter data by column name with dots', () => {
    const expectedData = [
      {
        name: 'Dav',
        age: 69,
        car: {
          make: 'Mitsubishi',
        },
      },
    ];

    RenderedMockedComponent.props.addFilter({
      columnName: 'car.make',
      value: 'Mitsubishi',
    });

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should remove filter', () => {
    const expectedData = cloneDeep(data);

    RenderedMockedComponent.props.addFilter({
      columnName: 'car.make',
      value: 'Mitsubishi',
    });

    RenderedMockedComponent.props.removeFilter({
      columnName: 'car.make',
    });

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should reset all filters', () => {
    const expectedData = cloneDeep(data);

    RenderedMockedComponent.props.addFilter({
      columnName: 'car.make',
      value: 'Mitsubishi',
    });

    RenderedMockedComponent.props.addFilter({
      columnName: 'name',
      value: 'Dav',
    });

    RenderedMockedComponent.props.resetFilter();

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should search all columns', () => {
    const expectedData = [
      {
        name: 'Caz',
        age: 49,
        car: {
          make: 'Land Rover',
        },
      },
      {
        name: 'Yule',
        age: 68,
        car: {
          make: 'Ford',
        },
      },
      {
        name: 'Ailsun',
        age: 41,
        car: {
          make: 'Honda',
        },
      },
    ];

    RenderedMockedComponent.props.handleSearchValue('o');

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should reset search value if search bar was closed', () => {
    const expectedData = cloneDeep(data);

    RenderedMockedComponent.props.toggleSearchBar();
    RenderedMockedComponent.props.handleSearchValue('o');
    RenderedMockedComponent.props.toggleSearchBar();

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should display rows by option perPageOption', () => {
    const expectedDisplayDataOnFirstPage = [
      {
        name: 'Caz',
        age: 49,
        car: {
          make: 'Land Rover',
        },
      },
      {
        name: 'Dav',
        age: 69,
        car: {
          make: 'Mitsubishi',
        },
      },
      {
        name: 'Yule',
        age: 68,
        car: {
          make: 'Ford',
        },
      },
      {
        name: 'Fredra',
        age: 67,
        car: {
          make: 'GMC',
        },
      },
      {
        name: 'Ursula',
        age: 80,
        car: {
          make: 'BMW',
        },
      },
    ];

    expect(RenderedMockedComponent.props.displayData).toEqual(
      expectedDisplayDataOnFirstPage
    );
  });

  it('should change count of displayed rows', () => {
    const expectedDisplayDataOnFirstPage = cloneDeep(data);

    RenderedMockedComponent.props.changePerPage(10);

    expect(RenderedMockedComponent.props.displayData).toEqual(
      expectedDisplayDataOnFirstPage
    );
  });

  it('should paginate', () => {
    const expectedDisplayDataOnSecondPage = [
      {
        name: 'Ailsun',
        age: 41,
        car: {
          make: 'Honda',
        },
      },
    ];

    RenderedMockedComponent.props.changePage(1);

    expect(RenderedMockedComponent.props.displayData).toEqual(
      expectedDisplayDataOnSecondPage
    );
  });

  it('should delete selected data', () => {
    const expectedData = [
      {
        name: 'Caz',
        age: 49,
        car: {
          make: 'Land Rover',
        },
      },
    ];

    RenderedMockedComponent.props.handleDelete([
      data[1],
      data[2],
      data[3],
      data[4],
      data[5],
    ]);

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should remove only deleted data items from selected data', () => {
    const expectedSelectedData = [data[0], data[1]];

    RenderedMockedComponent.props.handleSelect([
      data[0],
      data[1],
      data[2],
      data[3],
    ]);
    RenderedMockedComponent.props.handleDelete([data[2], data[3]]);

    expect(RenderedMockedComponent.props.selectedData).toEqual(
      expectedSelectedData
    );
  });

  it('should delete from raw data', () => {
    const expectedData = [
      {
        name: 'Dav',
        age: 69,
        car: {
          make: 'Mitsubishi',
        },
      },
      {
        name: 'Yule',
        age: 68,
        car: {
          make: 'Ford',
        },
      },
      {
        name: 'Fredra',
        age: 67,
        car: {
          make: 'GMC',
        },
      },
      {
        name: 'Ursula',
        age: 80,
        car: {
          make: 'BMW',
        },
      },
      {
        name: 'Ailsun',
        age: 41,
        car: {
          make: 'Honda',
        },
      },
    ];

    RenderedMockedComponent.props.handleSearchValue('a');
    RenderedMockedComponent.props.handleDelete([data[0]]);
    RenderedMockedComponent.props.handleSearchValue('');

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should select enumarated data', () => {
    const expectedSelectedData = [data[1], data[3], data[5]];

    RenderedMockedComponent.props.handleSelect([data[1], data[3], data[5]]);

    expect(RenderedMockedComponent.props.selectedData).toEqual(
      expectedSelectedData
    );
  });

  it('should select all data if there are no selected rows', () => {
    const expectedSelectedData = [
      data[0],
      data[1],
      data[2],
      data[3],
      data[4],
      data[5],
    ];

    RenderedMockedComponent.props.handleSelect([]);
    RenderedMockedComponent.props.toggleSelectAll();

    expect(RenderedMockedComponent.props.selectedData).toEqual(
      expectedSelectedData
    );
  });

  it('should deselect all data if there is at least one selected data item', () => {
    const expectedSelectedData = [];

    RenderedMockedComponent.props.handleSelect([data[4]]);
    RenderedMockedComponent.props.toggleSelectAll();

    expect(RenderedMockedComponent.props.selectedData).toEqual(
      expectedSelectedData
    );
  });

  it('should select data item if it was not selected', () => {
    const expectedSelectedData = [data[0]];

    RenderedMockedComponent.props.toggleSelectRow(data[0]);

    expect(RenderedMockedComponent.props.selectedData).toEqual(
      expectedSelectedData
    );
  });

  it('should deselect data item if it was selected before', () => {
    const expectedSelectedData = [];

    RenderedMockedComponent.props.handleSelect([data[0]]);
    RenderedMockedComponent.props.toggleSelectRow(data[0]);

    expect(RenderedMockedComponent.props.selectedData).toEqual(
      expectedSelectedData
    );
  });

  it('should construct filter lists for each column and sort lists', () => {
    const expectedFilterLists = {
      age: {
        label: 'Age',
        list: [41, 49, 67, 68, 69, 80],
      },
      'car.make': {
        label: 'Car make',
        list: ['BMW', 'Ford', 'GMC', 'Honda', 'Land Rover', 'Mitsubishi'],
      },
      name: {
        label: 'Name',
        list: ['Ailsun', 'Caz', 'Dav', 'Fredra', 'Ursula', 'Yule'],
      },
    };

    expect(RenderedMockedComponent.props.filterLists).toEqual(
      expectedFilterLists
    );
  });

  it('should exclude column from filter lists if option filterable is false', () => {
    const columns = [
      { name: 'name', label: 'Name', filterable: false },
      { name: 'age', label: 'Age', filterable: false },
      { name: 'car.make', label: 'Car make' },
    ];

    const expectedFilterLists = {
      'car.make': {
        label: 'Car make',
        list: ['BMW', 'Ford', 'GMC', 'Honda', 'Land Rover', 'Mitsubishi'],
      },
    };

    const RenderedMockedComponent = renderer
      .create(<ReactMUIDatatable data={data} columns={columns} />)
      .root.findByType(MockedComponent);

    expect(RenderedMockedComponent.props.filterLists).toEqual(
      expectedFilterLists
    );
  });

  it('should exclude column from search, if option searchable is false', () => {
    const columns = [
      { name: 'name', label: 'Name', searchable: false },
      { name: 'age', label: 'Age' },
      { name: 'car.make', label: 'Car make' },
    ];

    const expectedData = [
      { age: 49, car: { make: 'Land Rover' }, name: 'Caz' },
    ];

    const RenderedMockedComponent = renderer
      .create(<ReactMUIDatatable data={data} columns={columns} />)
      .root.findByType(MockedComponent);

    RenderedMockedComponent.props.handleSearchValue('e');

    expect(RenderedMockedComponent.props.computedData).toEqual(expectedData);
  });

  it('should return default toolbarSelectActions with delete icon if this prop was not received', () => {
    const ToolbarSelectActions = RenderedMockedComponent.props.toolbarSelectActions(
      {}
    );

    const tree = renderer.create(ToolbarSelectActions);

    expect(tree).toMatchSnapshot();
  });

  it('should return string for selectedData localization', () => {
    expect(
      RenderedMockedComponent.props.localization.toolbarSelect.selectedData(2)
    ).toBe('2 row(s) selected');
  });

  it('should return string for displayedRows localization', () => {
    expect(
      RenderedMockedComponent.props.localization.pagination.displayedRows({
        from: 1,
        to: 5,
        count: 10,
      })
    ).toBe('1-5 of 10');
  });

  it('it should fire onStateChanged event if one of state properties was changed', () => {
    /**
     *  Used directly "setState" because we need to check only triggering `onStateChanged`
     */
    RenderedMockedComponent.props.setShowSearchBar(true);
    RenderedMockedComponent.props.setSearchValue('Caz');
    RenderedMockedComponent.props.setSort({
      columnName: 'firstName',
      direction: 'DESC',
    });
    RenderedMockedComponent.props.setFilterValues({ firstName: 'Caz' });
    RenderedMockedComponent.props.setPage(1);
    RenderedMockedComponent.props.setPerPage(15);
    RenderedMockedComponent.props.setSelectedData([data[0]]);

    expect(onStateChanged).toBeCalledTimes(7);
    expect(onStateChanged).toHaveBeenLastCalledWith({
      name: 'selectedData',
      value: [data[0]],
      state: {
        searchValue: 'Caz',
        showSearchBar: true,
        sort: { columnName: 'firstName', direction: 'DESC' },
        filterValues: { firstName: 'Caz' },
        page: 1,
        perPage: 15,
        selectedData: [data[0]],
      },
    });
  });
});
