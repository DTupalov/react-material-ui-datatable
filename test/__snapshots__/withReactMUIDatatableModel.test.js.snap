// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withReactMUIDatatableModel should return default toolbarSelectActions with delete icon if this prop was not received 1`] = `
<button
  className="MuiButtonBase-root-7 MuiIconButton-root-1"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiIconButton-label-6"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-10"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </span>
  <span
    className="MuiTouchRipple-root-19"
  />
</button>
`;
